direct/transaction-manager/src/test/java/cz/o2/proxima/direct/transaction/manager/TransactionIT.java

```diff
@@ -233,27 +238,73 @@ public void testDeletedAttributeGet() throws InterruptedException {
     verifyNumInAttributeIs(key, numWrites + 1, attrA);
   }
 
+  @Test(timeout = 100_000)
+  public void testDeletedAttributeGetWithFailedWrite() throws InterruptedException {
+    // check atomic swap of data between two attributes
+    // a value is read from attribute X, incremented and written to attribute Y and deleted from X
+    // if value is not present in attribute X, it is read from attribute Y, and written to X
+
+    int numWrites = 300;
+    int numThreads = 10;
+
+    CountDownLatch latch = new CountDownLatch(numThreads);
+    ExecutorService service = direct.getContext().getExecutorService();
+    AtomicReference<Throwable> err = new AtomicReference<>();
+    int numWritesPerThread = numWrites / numThreads;
+    String attrA = device.toAttributePrefix() + "A";
+    String attrB = device.toAttributePrefix() + "B";
+    String key = "key";
+    // seed value
+    writeSeedValue(attrA, key);
+    for (int i = 0; i < numThreads; i++) {
+      service.submit(
+          () -> {
+            try {
+              for (int j = 0; j < numWritesPerThread; j++) {
+                if (!swapValueBetween(key, attrA, attrB, true)) {
+                  // failed write, repeat
+                  j--;
+                }
+              }
+            } catch (InterruptedException ex) {
+              Thread.currentThread().interrupt();
+            } catch (Throwable ex) {
+              err.set(ex);
+            }
+            latch.countDown();
+          });
+    }
+
+    latch.await();
+    if (err.get() != null) {
+      throw new RuntimeException(err.get());
+    }
+    verifyNumInAttributeIs(key, numWrites + 1, attrA);
+  }
+
   private void writeSeedValue(String attribute, String key) {
-    String transactionId = "firstTransaction";
-    BlockingQueue<Long> sequenceId = new ArrayBlockingQueue<>(1);
-    client.begin(
-        transactionId,
-        (a, b) -> ExceptionUtils.unchecked(() -> sequenceId.put(b.getSeqId())),
-        Collections.singletonList(
-            KeyAttributes.ofMissingAttribute(user, key, device, device.extractSuffix(attribute))));
-    StreamElement upsert =
-        device.upsert(
-            ExceptionUtils.uncheckedFactory(sequenceId::take),
-            key,
-            device.extractSuffix(attribute),
-            System.currentTimeMillis(),
-            ByteBuffer.allocate(4).putInt(1).array());
-    client.commit(transactionId, Collections.singletonList(KeyAttributes.ofStreamElement(upsert)));
-    view.write(upsert, (succ, exc) -> {});
+    try (OnlineAttributeWriter writer = Optionals.get(direct.getWriter(device))) {
+      StreamElement upsert =
+          device.upsert(
+              key,
+              device.extractSuffix(attribute),
+              System.currentTimeMillis(),
+              ByteBuffer.allocate(4).putInt(1).array());
+      CountDownLatch latch = new CountDownLatch(1);
+      writer.write(upsert, (succ, exc) -> latch.countDown());
+      ExceptionUtils.unchecked(latch::await);
```

Comment:
I think, we should always try to specify timeouts for awaits, maybe we can have a helper method to make this more convenient, eg..

```java
  public static void await(CountDownLatch latch, long timeout, TimeUnit unit) {
    try {
      assertTrue(latch.await(timeout, unit));
    } catch (InterruptedException e) {
      throw new IllegalStateException("Interrupted while waiting for latch to count down.");
    }
  }
  ```